src/xform_to_datalog/datalog_facts.cc

```diff
@@ -0,0 +1,149 @@
+  for (const arcs::ParticleSpecProto &particle_spec_proto :
+    manifest_proto.particle_specs()) {
+    amt::ParticleSpec particle_spec =
+        amt::ParticleSpec::CreateFromProto(particle_spec_proto);
+    std::string particle_spec_name = particle_spec.name();
+    auto ins_res = particle_specs.insert({
+      std::move(particle_spec_name), std::move (particle_spec) });
+    CHECK(ins_res.second) << "Name collision on particle spec.";
+  }
+
+  const std::string generated_recipe_name_prefix = "GENERATED_RECIPE_NAME";
+  uint64_t recipe_num = 0;
+  for (const arcs::RecipeProto &recipe_proto : manifest_proto.recipes()) {
+    const std::string &recipe_name =
+        recipe_proto.name().empty()
+          ? absl::StrCat(generated_recipe_name_prefix, recipe_num++)
+          : recipe_proto.name();
+
+    uint64_t particle_num = 0;
+    for (const arcs::ParticleProto &particle_proto : recipe_proto.particles()) {
+      const std::string &particle_spec_name = particle_proto.spec_name();
+      CHECK(!particle_spec_name.empty())
+        << "Particle with empty spec_name field not allowed.";
+      const std::string particle_name =
+          absl::StrCat(particle_spec_name, "#", particle_num++);
+
+      absl::flat_hash_map<ir::AccessPathRoot, ir::AccessPathRoot>
+          instantiation_map;
+      for (const arcs::HandleConnectionProto &connection_proto :
+        particle_proto.connections()) {
+        std::string handle_spec_name = connection_proto.name();
+        CHECK(!handle_spec_name.empty())
+          << "Handle connection with empty name field not allowed.";
+        std::string handle_name = connection_proto.handle();
+        CHECK(!handle_name.empty())
+          << "Handle with empty handle field not allowed.";
+
+        raksha::ir::HandleConnectionSpecAccessPathRoot spec_handle_root(
+            particle_spec_name, handle_spec_name);
+        raksha::ir::HandleConnectionAccessPathRoot
+          instantiated_handle_root(
+              recipe_name, particle_name, std::move(handle_spec_name));
+        raksha::ir::HandleAccessPathRoot handle_root(
+            recipe_name, std::move(handle_name));
+        CHECK(connection_proto.has_type())
+          << "Handle connection with absent type not allowed.";
+        std::unique_ptr<types::Type> connection_type =
+            types::Type::CreateFromProto(connection_proto.type());
+        ir::AccessPathSelectorsSet access_path_selectors_set =
+            connection_type->GetAccessPathSelectorsSet();
+
+        // For each Selector path from a root to a leaf, draw dataflow edges
+        // to and from the AccessPaths rooted at the handle at at the
+        // particle's handle connection. In the case that the handle
+        // connection is input only, the edge out will be unreachable from
+        // the rest of the graph. In the case that the connection is output
+        // only, the edge in will be unreachable from the rest of the graph.
```

Comment:
Not sure I understand these comments. Why would they be unreachable?

Reply:
Because of how we construct the edges within the `ParticleSpec`. Within the `ParticleSpec`, where the `HandleConnectionSpec` is directly accessible, we shall draw edges that start at a particular `HandleConnectionSpec` only if that `HandleConnectionSpec` is `read`, and we will draw edges that end at a `HandleConnectionSpec` only if that `HandleConnectionSpec` is `write`. That means that we can draw bidirectional edges between a `Handle` and a `HandleConnection` and rely upon the edge construction within a `ParticleSpec` to effectively filter the unnecessary input or output edge.

Reply:
I've updated the comment here; let me know if you buy my argument.

Reply:
As we discussed over video, while my reasoning is correct, it's not at all trivial to prove it and it produces more runtime overhead to produce these extra edges. I'm going to fix this to only produce the edges required.

Reply:
Done.